### PR TITLE
implements proposed max-props rule

### PR DESCRIPTION
**Pull request by kevsommer**
This is a WIP PR that addresses issue #2429. 

It implements the rule for both, the Composition and Options API.

However, I haven't been able to get the case for `defineProps` with type-only declaration to work.

```vue
<script setup lang="ts">
defineProps<{ prop1: string, prop2: string, }>();
</script>
```

### CODE REVIEW

**Comment by kevsommer** [comment]
Should work for `defineProps` with type-only declaration now.

File: docs/rules/max-props.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  prop1: String
+})
+</script>
+```
+
+</eslint-code-block>
+
+Nothing.
+
+## :mag: Implementation
```
Comment (FloEdelmann) [inline docs/rules/max-props.md:61]: "Nothing." is only the placeholder for when there are no rule options.

```suggestion
</eslint-code-block>

## :mag: Implementation
```

File: lib/rules/max-props.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          maxProps: {
+            type: 'integer',
+            minimum: 1
+          }
+        },
+        additionalProperties: false
```
Comment (FloEdelmann) [inline lib/rules/max-props.js:26]:
```suggestion
        additionalProperties: false,
        minProperties: 1
```

File: lib/rules/max-props.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    messages: {
+      tooManyProps:
+        'Component has too many props ({{propCount}}). Maximum allowed is {{limit}}.'
+    }
+  },
+  /** @param {RuleContext} context */
+  create(context) {
+    /**
+     * @type {Record<string, number>}
+     */
```
Comment (FloEdelmann) [inline lib/rules/max-props.js:38]:
```suggestion
    /** @type {Record<string, number>} */
```

File: lib/rules/max-props.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    const option = context.options[0] || {}
+
+    /**
+     *
+     * @param {import('../utils').ComponentProp[]} props
+     */
```
Comment (FloEdelmann) [inline lib/rules/max-props.js:44]:
```suggestion
    /**
     * @param {import('../utils').ComponentProp[]} props
     */
```

File: tests/lib/rules/max-props.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+})
+
+tester.run('max-props', rule, {
+  valid: [
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+
+      </template>
```
Comment (FloEdelmann) [inline tests/lib/rules/max-props.js:25]: Please remove the empty `<template>` blocks from the test cases, they don't add any value.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Thanks, this looks quite good! :slightly_smiling_face: 
I only have some minor nitpicks.

**Comment by kevsommer** [comment]
Thank you for the review! I have implemented the requested changes :)

**Review by FloEdelmann — APPROVED** [review]
Looks good to me :slightly_smiling_face:

**Review by ota-meshi — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**